### PR TITLE
Adding, shifting, and stacking for TextDiagramDrawer

### PR DESCRIPTION
**Pull request by bryano**

### CODE REVIEW

File: cirq/circuits/text_diagram_drawer.py [hunk 1]
```diff
@@ -243,6 +260,73 @@ def render(self,
...
+        shifted_vertical_padding = {y + dy: padding
+                for y, padding in self.vertical_padding.items()}
+        return self.__class__(
+                entries=shifted_entries,
+                horizontal_lines=shifted_horizontal_lines,
+                vertical_lines=shifted_vertical_lines,
+                horizontal_padding=shifted_horizontal_padding,
+                vertical_padding=shifted_vertical_padding)
+
+    def __iadd__(self, other: 'TextDiagramDrawer') -> 'TextDiagramDrawer':
```
Comment (Strilanc) [inline cirq/circuits/text_diagram_drawer.py]: I don't think that this is there is a natural unambiguous meaning for "adding" two diagrams. E.g. I would have expected addition to do horizontal concatenation, as it does for strings and lists in python. I think we should give this a name instead of an operator e.g. "merge".

File: cirq/circuits/text_diagram_drawer.py [hunk 2]
```diff
@@ -243,6 +260,73 @@ def render(self,
...
+        self.vertical_padding.update(other.vertical_padding)
+        return self
+
+    def __add__(self, other: 'TextDiagramDrawer') -> 'TextDiagramDrawer':
+        summed = self.copy()
+        summed += other
+        return summed
+
+    @staticmethod
+    def vstack(top: 'TextDiagramDrawer', bottom: 'TextDiagramDrawer',
```
Comment (Strilanc) [inline cirq/circuits/text_diagram_drawer.py]: Needs a docstring.

Why is there a 'padding resolver' argument?

Comment (Strilanc) [inline cirq/circuits/text_diagram_drawer.py]: In numpy vstack takes an iterable of tensors and merges them into one tensor. Shouldn't this then take a list of diagrams, instead of two separate diagram arguments?

Comment (bryano) [reply cirq/circuits/text_diagram_drawer.py]: To determine what the padding should be when the diagrams to be vertically stacked have different values for horizontal padding in the same column.

Comment (bryano) [reply cirq/circuits/text_diagram_drawer.py]: Done.

Comment (bryano) [reply cirq/circuits/text_diagram_drawer.py]: Docstring added.

File: cirq/circuits/text_diagram_drawer.py (hunk 1 above)
Comment (bryano) [reply cirq/circuits/text_diagram_drawer.py]: Changed to "update".

**Comment by bryano** [comment]
Do you see where the circular import is being introduced?

**Comment by bryano** [comment]
I'm tempted to remove the use of `value_equality` to get this passing the tests, but it'd be nice to have it in and I'd like to understand why the tests are failing.

File: cirq/circuits/text_diagram_drawer.py [hunk 3]
```diff
@@ -243,6 +262,90 @@ def render(self,
...
+
+    @classmethod
+    def vstack(cls,
+               diagrams: Iterable['TextDiagramDrawer'],
+               padding_resolver: Callable[[Sequence[int]], int]=max):
+        """Vertically stack text diagrams.
+
+        Args:
+            diagrams: The diagrams to stack, ordered from bottom to top.
+            padding_resolver: A function that takes a list of paddings
```
[the remark this answers is not in the document]
Comment (bryano) [reply cirq/circuits/text_diagram_drawer.py]: Consider the following:
```
d = TextDiagramDrawer()
d.write(0, 0, 'a')
d.write(1, 0, 'b')
d1, d2, d3, d4 = (d.copy() for _ in (1, 2, 3, 4))
d2.force_horizontal_padding_after(0, 3)
d3.force_horizontal_padding_after(0, 5)
d4.force_horizontal_padding_after(0, 2)
stacked = TextDiagramDrawer.vstack((d1, d2, d3, d4))
```
What should `stacked.horizontal_padding[0]` be? The smallest padding specified (2) or the largest (5)? The first (3) or the last (2)? The rounded mean? I imagine the answer will be context dependent, so I allowed the caller to specify how to resolve potential discrepancies.

Both `max` and `min` seem like reasonable defaults to me. An alternative default is to through an error  if different paddings are specified, in order to force the caller to say explicitly how to resolve them.

[the remark this answers is not in the document]
Comment (Strilanc) [reply cirq/circuits/text_diagram_drawer.py]: Oh I see, the issue is that columns may have incompatible padding requirements.

I would default the padding resolver to `None` and raise an error when conflicts occurs. A conflict occurs if the set of horizontal paddings for any column (with None meaning no forced padding) has a cardinality greater than 1.

[the remark this answers is not in the document]
Comment (bryano) [reply cirq/circuits/text_diagram_drawer.py]: To clarify, if one or more diagrams specify the same padding for a given column and one or more diagrams don't specify any padding for that column, should an error be thrown?

[the remark this answers is not in the document]
Comment (Strilanc) [reply cirq/circuits/text_diagram_drawer.py]: Correct.

[the remark this answers is not in the document]
Comment (bryano) [reply cirq/circuits/text_diagram_drawer.py]: Done.

**Comment by Strilanc** [comment]
Conflicts with the text diagram changes I made last week. Needs to be updated.

**Comment by bryano** [comment]
Done

File: cirq/circuits/text_diagram_drawer.py [hunk 4]
```diff
@@ -305,3 +331,126 @@ def render(self,
...
+def _same_element_or_throw_error(elements: Sequence[Any]):
+    """Extract an element or throw an error.
+
+    Args:
+        elements: A sequence of something.
+
+    copies of it. Returns None on an empty sequence.
+
+    Raises:
+        ValueError: The sequence constai
```
Comment (Strilanc) [inline cirq/circuits/text_diagram_drawer.py]: constai?

File: cirq/circuits/text_diagram_drawer_test.py [hunk 5]
```diff
@@ -15,8 +15,40 @@
...
+
+    Args:
+        actual: The text diagram.
+        desired: The desired rendering as a string.
+        **kwargs: Keyword arguments to be passed to actual.render.
+    """
+    actual_diagram = actual.render(**kwargs)
+    desired_diagram = desired
+    assert actual_diagram == desired_diagram, (
+    "Diagram's rendering differs from the desired rendering.\n"
```
Comment (Strilanc) [inline cirq/circuits/text_diagram_drawer_test.py]: Nit: indent

File: cirq/circuits/text_diagram_drawer_test.py [hunk 6]
```diff
@@ -159,3 +191,165 @@ def test_multiline_entries():
...
+    hstacked_min = TextDiagramDrawer.hstack((d, dd), padding_resolver=min)
+    expected = """
+AB D
+
+ C E  F
+    """.strip()
+    assert_has_rendering(hstacked_min, expected)
+
+def test_drawer_eq():
+    assert TextDiagramDrawer().__eq__(23) == NotImplemented
```
Comment (Strilanc) [inline cirq/circuits/text_diagram_drawer_test.py]: Use `EqualsTester` to properly exercise the eq, neq, etc methods.

File: cirq/circuits/text_diagram_drawer.py [hunk 7]
```diff
@@ -59,12 +62,35 @@ class TextDiagramDrawer:
...
+                else dict(horizontal_padding)
+                ) # type: Dict[int, Union[int, float]]
+        self.vertical_padding = (dict() if vertical_padding is None
+                else dict(vertical_padding)
+                ) # type: Dict[int, Union[int, float]]
+
+    def __eq__(self, other):
+        if not isinstance(other, type(self)):
+            return NotImplemented
+        attrs = ('entries', 'horizontal_lines', 'vertical_lines',
```
Comment (Strilanc) [inline cirq/circuits/text_diagram_drawer.py]: Looks like this is a prime candidate for `@cirq.value_equality` (well, `@value.value_equality`).

File: cirq/circuits/text_diagram_drawer_test.py (hunk 6 above)
Comment (bryano) [reply cirq/circuits/text_diagram_drawer_test.py]: Done

File: cirq/circuits/text_diagram_drawer.py (hunk 4 above)
Comment (bryano) [reply cirq/circuits/text_diagram_drawer.py]: fixed

File: cirq/circuits/text_diagram_drawer.py (hunk 7 above)
Comment (bryano) [reply cirq/circuits/text_diagram_drawer.py]: I completely agree, but when I try to use that I get circular import errors that I don't know how to fix.

Comment (Strilanc) [reply cirq/circuits/text_diagram_drawer.py]: Might be a cycle via `cirq.protocols`... What's the stack trace?

File: cirq/circuits/text_diagram_drawer.py [hunk 8]
```diff
@@ -83,14 +83,15 @@ def __init__(self,
                 else dict(vertical_padding)
                 ) # type: Dict[int, Union[int, float]]
 
+    def _value_equality_values_(self):
+        attrs = ('entries', 'horizontal_lines', 'vertical_lines',
+                 'horizontal_padding', 'vertical_padding')
+        return tuple(getattr(self, attr) for attr in attrs)
+
     def __eq__(self, other):
```
Comment (Strilanc) [inline cirq/circuits/text_diagram_drawer.py]: Delete the `__eq__` method when using value equality.

File: cirq/circuits/text_diagram_drawer.py (hunk 7 above)
Comment (bryano) [reply cirq/circuits/text_diagram_drawer.py]: ../../.virtualenvs/quantumlib/lib/python3.6/site-packages/py/_path/local.py:668: in pyimport
    __import__(modname)
cirq/__init__.py:21: in <module>
    from cirq.circuits import (
cirq/circuits/__init__.py:17: in <module>
    from cirq.circuits.text_diagram_drawer import (
cirq/circuits/text_diagram_drawer.py:21: in <module>
    from cirq import value
cirq/value/__init__.py:26: in <module>
    from cirq.value.periodic_value import (
cirq/value/periodic_value.py:19: in <module>
    import cirq.protocols
cirq/protocols/__init__.py:16: in <module>
    from cirq.protocols.apply_unitary import (
cirq/protocols/apply_unitary.py:23: in <module>
    from cirq import linalg
cirq/linalg/__init__.py:28: in <module>
    from cirq.linalg.decompositions import (
cirq/linalg/decompositions.py:267: in <module>
    @value.value_equality
E   AttributeError: module 'cirq.value' has no attribute 'value_equality'

File: cirq/circuits/text_diagram_drawer.py (hunk 8 above)
Comment (bryano) [reply cirq/circuits/text_diagram_drawer.py]: Yea, that's just temporary because the class isn't actually dressed with `value_equality` yet.

Comment (Strilanc) [reply cirq/circuits/text_diagram_drawer.py]: We need to resolve the dependency cycle so it can be annotated with it.

Comment (bryano) [reply cirq/circuits/text_diagram_drawer.py]: Done

**Comment by bryano** [comment]
@Strilanc Tests are only failing in pytest2. Is that worth fixing given that Python 2 support is going to be removed?

**Comment by vtomole** [comment]
@bryano Depends on if you want these features in the next release as we are removing Python 2 after it.

**Comment by bryano** [comment]
I think I'll wait it out. pytest2 passes on my machine, so I can't even replicate the failures to address them.

**Comment by bryano** [comment]
@Strilanc Tests passing now that pytest2 is gone.

**Comment by vtomole** [comment]
@bryano @Strilanc Checks are passing with removal of python 2. Time to merge?

**Comment by vtomole** [comment]
Related to: https://github.com/quantumlib/Cirq/issues/1226

**Comment by bryano** [comment]
@Strilanc I noticed that fsim_gate.py imports cirq (rather than its submodules). Is there a reason for that?

**Comment by bryano** [comment]
Also, what do you think about renaming either the file 'value_equality.py' or the decorator 'value_equality'. I think a lot of the import difficulties are caused by there being a module and method of the same name.

**Comment by vtomole** [comment]
@Strilanc  Looks like https://travis-ci.com/quantumlib/Cirq/jobs/210593629 format is running into code that wasn't touched by this pull request.

File: cirq/circuits/text_diagram_drawer.py [hunk 9]
```diff
@@ -305,3 +330,124 @@ def render(self,
...
+                              horizontal_padding=self.horizontal_padding)
+
+    def shift(self, dx: int = 0, dy: int = 0) -> 'TextDiagramDrawer':
+        self._transform_coordinates(lambda x, y: (x + dx, y + dy))
+        return self
+
+    def shifted(self, dx: int = 0, dy: int = 0) -> 'TextDiagramDrawer':
+        return self.copy().shift(dx, dy)
+
+    def update(self, other: 'TextDiagramDrawer') -> 'TextDiagramDrawer':
```
Comment (Strilanc) [inline cirq/circuits/text_diagram_drawer.py]: This method needs a better name. It overwrites entries but adds lines? Looks like `draw_on_top_of` or similar.

File: cirq/linalg/decompositions.py [hunk 10]
```diff
@@ -329,8 +329,7 @@ def canonicalize(self, atol: float = 1e-8) -> 'AxisAngleDecomposition':
 
     def _value_equality_values_(self):
         v = self.canonicalize(atol=0)
-        return (value.PeriodicValue(v.angle,
-                                    period=math.pi * 2), v.axis, v.global_phase)
+        return (v.angle % (math.pi * 2), v.axis, v.global_phase)
```
Comment (Strilanc) [inline cirq/linalg/decompositions.py]: This is an incorrect change! It breaks the approximate comparison logic.

File: cirq/ops/common_channels.py [hunk 11]
```diff
@@ -548,7 +549,7 @@ def phase_damp(gamma: float) -> PhaseDampingChannel:
     return PhaseDampingChannel(gamma)
 
 
-@value.value_equality
+@value_equality
```
Comment (Strilanc) [inline cirq/ops/common_channels.py]: This was really the only way to break the cycle? It's quite odd... we really don't want to have packages outside of a tree digging into the guts of that tree.

File: cirq/contrib/acquaintance/permutation.py [hunk 12]
```diff
@@ -112,14 +114,14 @@ class LinearPermutationGate(PermutationGate):
     def __init__(self,
                  num_qubits: int,
                  permutation: Dict[int, int],
-                 swap_gate: ops.Gate=ops.SWAP
-                 ) -> None:
+                 swap_gate: Optional[ops.Gate] = None) -> None:
```
Comment (Strilanc) [inline cirq/contrib/acquaintance/permutation.py]: I've fixed the bug in the formatter where it would compare against a too-recent version of master. You should be able to revert almosta ll the formatting changes.

File: cirq/contrib/acquaintance/permutation.py [hunk 13]
```diff
@@ -89,7 +90,8 @@ def _circuit_diagram_info_(self, args: protocols.CircuitDiagramInfoArgs
 class SwapPermutationGate(PermutationGate):
     """Generic swap gate."""
 
-    def __init__(self, swap_gate: ops.Gate=ops.SWAP):
+    def __init__(self, swap_gate: Optional[ops.Gate] = None):
+        swap_gate = ops.SWAP if swap_gate is None else swap_gate
```
Comment (Strilanc) [inline cirq/contrib/acquaintance/permutation.py]: This functional change is not related to the PR.

File: cirq/contrib/acquaintance/permutation.py [hunk 14]
```diff
@@ -35,9 +35,10 @@ class PermutationGate(ops.Gate, metaclass=abc.ABCMeta):
             qubits (e.g. SWAP or fermionic swap).
     """
 
-    def __init__(self, num_qubits: int, swap_gate: ops.Gate=ops.SWAP) -> None:
+    def __init__(self, num_qubits: int,
+                 swap_gate: Optional[ops.Gate] = None) -> None:
         self._num_qubits = num_qubits
-        self.swap_gate = swap_gate
+        self.swap_gate = ops.SWAP if swap_gate is None else swap_gate
```
Comment (Strilanc) [inline cirq/contrib/acquaintance/permutation.py]: This functional change is not related to the PR.

File: cirq/ops/common_channels.py (hunk 11 above)
Comment (bryano) [reply cirq/ops/common_channels.py]: Better now?

File: cirq/linalg/decompositions.py (hunk 10 above)
Comment (bryano) [reply cirq/linalg/decompositions.py]: New cycle avoidance technique makes this irrelevant.

File: cirq/circuits/text_diagram_drawer.py (hunk 9 above)
Comment (bryano) [reply cirq/circuits/text_diagram_drawer.py]: How about `superimpose`? In any case, it would be good for the name to indicate (as I think `update` does) that the changes are made in place. So maybe `superimpose_inplace` or `inplace_superimpose`?

**Comment by bryano** [comment]
Okay, I think at this point the only two outstanding items are the cycle-avoidance and the `update` renaming.

File: cirq/contrib/acquaintance/permutation.py (hunk 13 above)
Comment (bryano) [reply cirq/contrib/acquaintance/permutation.py]: This was a hack to avoid a mypy complaint that doesn't seem to be happening anymore.

File: cirq/circuits/text_diagram_drawer.py (hunk 9 above)
Comment (Strilanc) [reply cirq/circuits/text_diagram_drawer.py]: Just `superimpose` is great. You can add `superimposed` as a variant that operates on a copy, similar to `shift` vs `shifted`.

**Comment by Strilanc** [comment]
I agree.

It looks like the cycle is much better with the small number of import changes you're making now.

File: cirq/ops/common_channels.py (hunk 11 above)
Comment (Strilanc) [reply cirq/ops/common_channels.py]: Yes

**Comment by bryano** [comment]
Great. Just let me know what to rename `update` and then it's good to go.

**Comment by Strilanc** [comment]
`superimpose` and a copy variant `superimposed`

**Comment by bryano** [comment]
Done

**Comment by bryano** [comment]
@Strilanc ping

**Comment by dabacon** [comment]
This seems to have grown stale.  I'm closing.  If it can be salvaged, please open a new PR.

**Comment by bryano** [comment]
I think this is ready to go and just awaiting final approval from @Strilanc

**Comment by dabacon** [comment]
@Strilanc this has been sitting on you for 18 days now, could you take a look?

**Comment by Strilanc** [comment]
Close as stale.

**Comment by bryano** [comment]
@Strilanc I thought this was ready to merge. I'll get it back to up to sync if it'll get reviewed.

**Comment by bryano** [comment]
@Strilanc PTAL

**Comment by bryano** [comment]
@Strilanc ping

**Comment by vtomole** [comment]
Does this close https://github.com/quantumlib/Cirq/issues/1226?

**Comment by bryano** [comment]
No, but #1240 was a start towards #1226.

**Comment by vtomole** [comment]
@bryano This should have been merged by now in my opinion. @Strilanc Is there a reason you are ambivalent toward this being in Cirq? I'll bring this up in the sync tommorrow.

**Comment by Strilanc** [comment]
Darn history size page limits.
